### PR TITLE
Fix pin_memory to return correctly typed sequences

### LINKED ISSUE #48419: pin_memory=True in DataLoader converts a tuple to list automatically
## 🐛 Bug

DataLoader with `pin_memory=True` converts the tuple returned from `Dataset` & `collate_fn` to a list automatically. This is because of this particular [line in the `pin_memory` implementation](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/_utils/pin_memory.py#L55)  . The actual type of the return item shouldn't be changed.

```python
elif isinstance(data, container_abcs.Sequence):
    return [pin_memory(sample) for sample in data]
```

## To Reproduce

```python
class TestDataset(IterableDataset):
    def __iter__(self):
        for i in range(30):
            yield torch.rand(3, 10), torch.rand(3, 3)

def collate(batch):
    return batch[0]

dset = TestDataset()

dloader = DataLoader(
    dataset=dset,
    batch_size=1,
    collate_fn=collate,
    num_workers=0,
    pin_memory=True
)

iterator = iter(dloader)
batch = next(iterator)
print(type(batch))   # prints: < class 'list' >


dloader = DataLoader(
    dataset=dset,
    batch_size=1,
    collate_fn=collate,
    num_workers=0,
    pin_memory=False)

iterator = iter(dloader)
batch = next(iterator)
print(type(batch))   # prints: < class 'tuple' >
```

## Expected behavior

The first `print` statement above should print `<class 'tuple'>`

## Environment

PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 19.10 (x86_64)
GCC version: (Ubuntu 9.2.1-9ubuntu2) 9.2.1 20191008
Clang version: Could not collect
CMake version: version 3.18.2

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 455.45.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] pytorch-lightning==1.0.5
[pip3] torch==1.7.0
[pip3] torchaudio==0.7.0
[conda] numpy                     1.19.4                   pypi_0    pypi
[conda] pytorch-lightning         1.0.5                    pypi_0    pypi
[conda] torch                     1.7.0                    pypi_0    pypi

### PR DESCRIPTION
**Pull request by scarecrow1123**
Fixes #48419 

cc: @SsnL

### CODE REVIEW

**Comment by SsnL** [comment]
Needs a test in test_dataloader.py

**Comment by mariosasko** [comment]
`iter` is redundant here.

This should do the job:
```
type(data)(pin_memory(sample) for sample in data)
```

File: torch/utils/data/_utils/pin_memory.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def pin_memory(data):
     elif isinstance(data, tuple) and hasattr(data, '_fields'):  # namedtuple
         return type(data)(*(pin_memory(sample) for sample in data))
     elif isinstance(data, container_abcs.Sequence):
-        return [pin_memory(sample) for sample in data]
+        return type(data)(iter(pin_memory(sample) for sample in data))
```
Comment (mariosasko) [inline torch/utils/data/_utils/pin_memory.py:55]: `iter` is redundant here

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Please add test and remove redundant `iter`

**Comment by VitalyFedyunin** [comment]
Also rebasing will clear onnx errors.

**Pull request closed without merging**